libsolutil/IpfsHash.cpp

```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
 	reverse(output.begin(), output.end());
 	return output;
 }
+struct Chunk
+{
+	Chunk() = default;
+	Chunk(bytes _hash, size_t _size, size_t _blockSize):
+		hash(std::move(_hash)),
+		size(_size),
+		blockSize(_blockSize),
+		links{},
+		lengths{}
+	{}
+	bytes hash = {};
+	size_t size = 0;
+	size_t blockSize = 0;
+
+	bytes links = {};
+	size_t linksBlockSize = 0;
+	bytes lengths = {};
+
+	void addLink(Chunk& _chunkLink)
+	{
+		size += _chunkLink.size;
+		linksBlockSize += _chunkLink.blockSize;
+
+		bytes newLinkData =
+			bytes {0x0a} +
+			varintEncoding(_chunkLink.hash.size()) +
+			std::move(_chunkLink.hash) +
+			bytes{0x12, 0x00, 0x18} + varintEncoding(_chunkLink.blockSize);
+
+		links += bytes{0x12} + varintEncoding(newLinkData.size()) + std::move(newLinkData);
+		lengths += bytes{0x20} + varintEncoding(_chunkLink.size);
+	}
+
+	void updateHash()
+	{
+		bytes tmp = bytes{0x08, 0x02, 0x18} + varintEncoding(size) + lengths;
+		bytes blockData = links + bytes{0x0a} + varintEncoding(tmp.size()) + tmp;
+
+		blockSize = blockData.size() + linksBlockSize;
+		hash = bytes{0x12, 0x20} + picosha2::hash256(blockData);
+	}
+};
+
+using Chunks = vector<Chunk>;
+
+Chunks buildNextLevel(Chunks& _currentLevel)
+{
+	size_t const maxChildNum = 174;
+
+	Chunks nextLevel;
+	Chunk newChunk;
+
+	for (size_t chunkIndex = 0; chunkIndex < _currentLevel.size(); chunkIndex++)
+	{
+		// new link that we are adding
+		newChunk.addLink(_currentLevel[chunkIndex]);
+
+		if (chunkIndex % maxChildNum == maxChildNum - 1 || chunkIndex == _currentLevel.size() - 1)
```

Comment:
If you collect the links in a local vector, you can use the vector's size for this check and do something like:
```
for (Chunk& chunk: _currentLevel)
{
  links.emplace_back(createLink(chunk));
  if (links.size() == maxChildNum - 1)
  {
    nextLevel.emplace_back(combineLinks(links));
    links = {};
  }
}
if (!links.empty())
    nextLevel.emplace_back(combineLinks(links));

return nextLevel;
}
```

Reply:
I created different structure to hold Link data and used local vector of links to store them